User-Guide Documentation Git as a Data Source (#184)
nautobot/nautobot: nautobot/docs/user-guides/git-data-source.md

# Git as a Data Source

> The Nautobot Git as a Data Source feature was developed to provide the ability to populate existing data, templates, scripts, and much more into Nautobot.  While using the benefits that Github already provides with issue tracking, discussions, pipelines, and approvals.

Reviewer: ```suggestion
Using Git as a Data Source feature was developed to provide the ability to populate existing data, templates, scripts, and much more into Nautobot; while leveraging the benefits that GitHub already provides including issue tracking, discussions, pipelines, and approvals.
```

> For more in depth details on how this feature [see](https://nautobot.readthedocs.io/en/latest/models/extras/gitrepository/).

Reviewer: Links like "see" or "here" are bad for SEO and should be avoided. Also, you accidentally a word.

Please change this link to be descriptive, like
```suggestion
For more technical details on how to use this feature, please see the documentation on [Git Repositories](https://nautobot.readthedocs.io/en/latest/models/extras/gitrepository/).
```

## Supported Providers
The feature uses the concept of a `provides` field to map a repository to a use case. A list of the supported options is provided below.

|Name|Summary|Related Plugin|
|:--|:--|:--|
|Export Templates|Nautobot allows users to define custom templates that can be used when exporting objects.|N/A|
|Jobs|Jobs are a way for users to execute custom logic on demand from within the Nautobot UI. Jobs can interact directly with Nautobot data to accomplish various data creation, modification, and validation tasks.|N/A|
|Config Contexts||[Golden Config](https://github.com/nautobot/nautobot-plugin-golden-config)|

> |Backup Configs|Backup configuration data.|[Golden Config](https://github.com/nautobot/nautobot-plugin-golden-config)|

Reviewer 2: I think you can remove these, imho, docs should only relate to core, but will defer to core committers.

Reviewer: I'm personally not opposed to listing the plugins, but I want to make sure it's done in a way that doesn't make it seem like our plugins are the only way to do it.

Author: Added `core functionality` to be more explicitly on the data sources that come natively.  If different terminology is wanted let me know.

Author: I've changed the table into two tables.

# Git as a Data Source

Using Git&trade; as a Data Source feature was developed to provide the ability to populate existing data, templates, scripts, and much more into Nautobot; while leveraging the benefits that GitHub already provides including issue tracking, discussions, pipelines, and approvals.

> For more technical details on how to use this feature, please see the documentation on [Git Repositories](https://nautobot.readthedocs.io/en/latest/models/extras/gitrepository/).

Reviewer 3: Relative rather than absolute link please.
```suggestion
For more technical details on how to use this feature, please see the documentation on [Git Repositories](../models/extras/gitrepository.md).
```

## Supported Providers
The feature uses the concept of a `provides` field to map a repository to a use case. A list of the supported options is provided below.

|Name|Summary|Related Plugin|
|:--|:--|:--|
|Export Templates|Nautobot allows users to define custom templates that can be used when exporting objects.|Core Functionality|
|Jobs|Jobs are a way for users to execute custom logic on demand from within the Nautobot UI. Jobs can interact directly with Nautobot data to accomplish various data creation, modification, and validation tasks.|Core Functionality|

> |Config Contexts|Config contexts can be used to provide additional data that you can't natively store in Nautobot.|Core Functionality|

Author: cross-links added.

|Backup Configs|Backup configuration data.|[Golden Config](https://github.com/nautobot/nautobot-plugin-golden-config)|
|Intended Configs|Stores the intended configurations, this grabs Nautobot data and runs through Jinja Templates.|[Golden Config](https://github.com/nautobot/nautobot-plugin-golden-config)|
|Jinja Templates|Repository that holds Jinja templates to be used to generate intended configs.|[Golden Config](https://github.com/nautobot/nautobot-plugin-golden-config)|

## Repository Details

Parameters:
|Field|Explanation|
|:--|:--|
|Name|User friendly name for the repo.|
|Slug|Auto-generated based on the `name` provided.|
|Remote URL|The URL pointing to the Git repo. Current git url usage is limited to `http` or `https`.|
|Branch|The branch in the Git repo to use. Defaults to `main`.|

> |Token|The token is a personal access token for the `username` provided.  For more information on generating a personal access token. [GitHub Personal Access Token](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)

Reviewer 3: We should be clear that this isn't GitHub-exclusive - much like the `nautobot/docs/models/extras/gitrepository.md`, this should probably include references to GitLab and Bitbucket as well?

> |Username|The Git username that corresponds with the personal access token above.|

Reviewer 3: Not required for GitHub, but required for GitLab

> |Provides|Valid providers for Git Repo.|

Reviewer 3: ```suggestion
|Provides|Resource type(s) provided by this Git repo.|
```

> <br>

Reviewer 3: Remove? Shouldn't be needed.

## Using Git Data Sources
This section will focus on examples and use the `user-guide` branch on the demo-git-datasources repo: `https://github.com/nautobot/demo-git-datasource`.

### Export Templates

> Export Templates allow a user to export Nautobot objects based on a custom template.  Export templates can change over time depending on the needs of a user.  Allowing export templates to reference a Git repo makes managing templates easier.

Reviewer 3: ```suggestion
[Export Templates](../models/extras/exporttemplate.md) allow a user to export Nautobot objects based on a custom template.  Export templates can change over time depending on the needs of a user.  Allowing export templates to reference a Git repo makes managing templates easier.
```

A template can be used to put objects into a specific format for ingestion into another system, tool, or report.  It is possible that different templates are needed depending on specific users or teams.  This can lead to sprawl of export templates.  To keep accurate templates synced with Nautobot the Git Data Sources extensibility feature can be used.

#### Step 1
Navigate to the Data Sources Git integration. **Extensibility -> Git Repositories**.

![](./images/git-data-source_1.png)

Click [+] or [Add]

That loads a default page to add a repository.

![](./images/git-data-source_2.png)

> > By default only config contexts, export templates, and jobs are implemented.  Other data sources will get added when a specific plugin is used.

Reviewer 3: ```suggestion
!!! note
    By default only config contexts, export templates, and jobs are available resource types.  Others may be added when specific plugins are used.
```

#### Step 2
Fill out the details for the Git repository. More information on the inputs can be found in the [fields section](#repository-details).

![](./images/git-data-source_3.png)

As soon as you click on **Create**, Nautobot will clone and sync the repository and provide status of the job.

![](./images/git-data-source_4.png)
![](./images/git-data-source_6.png)

The repository will now be displayed on the main Git Repository page.

![](./images/git-data-source_5.png)

Once the repository is synced each template will now be available in the Export Templates section.  **Extensibility -> Export Templates**.

![](./images/git-data-source_7.png)

> > Note: If the templates don't populate make sure the Git directory is named `export_templates` and the Nautobot `content type` for the object is the sub-directory name.

Reviewer 3: ```suggestion
> !!! note
    If the templates don't populate, make sure the Git directory is named `export_templates` and the sub-directory and sub-sub-directory names correctly match the Nautobot `content type`.
```

The power of having export templates utilizing the Git integration comes with the native source control features that Git comes with.  To illustrate a simple Git sync within Nautobot assume the following template needs to be updated.

Filename: `/export_templates/dcim/device/yaml_export.yml`

Current contents:
```jinja
---
{% for device in queryset %}
{% if device.status %}
- {{ device.name }}:
{% endif %}
{% endfor %}
...
```

The template needs to be modified to provide more information than just a list of hostnames.  The site needs to be added.

The updated template is now:
```jinja
---
{% for device in queryset %}
{% if device.status %}
- {{ device.name }}:
  site: {{ device.site }}
{% endif %}
{% endfor %}
...
```

Once the contributor updates the Git repository via normal Git processes and it is reviewed and merged into the branch that was used, a sync process from Nautobot needs to be completed.  This can be done from the default Git view, or within a specific detailed view of a Git repository.

From the default Git repositories view:
![](./images/git-data-source_8.png)

From the detailed view:
![](./images/git-data-source_9.png)

> > Once the repository has been synced its easy to check the history for the templates.
> Navigate to **Git Repositories** and select the repository in question.  Once you're in the detailed view you can look through the **Synchronization Status** or **Change Log** tabs.

Reviewer 3: ```suggestion
!!! tip
    Once the repository has been synced it's easy to check the history for the templates.
    Navigate to **Git Repositories** and select the repository in question.  Once you're in the detailed view you can look through the **Synchronization Status** or **Change Log** tabs.
```

Now that the Git repository is linked for export templates it can be controlled via the normal Git operations workflow, which allows users or groups of users to perform code reviews using Pull Requests etc.

### Jobs
Jobs are a way for users to execute custom logic on demand from within the Nautobot UI. Jobs can interact directly with Nautobot data to accomplish various data creation, modification, and validation tasks.

> For technical details on jobs, please see the documentation on [jobs](https://nautobot.readthedocs.io/en/latest/additional-features/jobs/#jobs).

Reviewer 3: ```suggestion
For technical details on jobs, please see the documentation on [jobs](../additional-features/jobs.md#jobs).
```

Jobs allow a user to write scripts in Python.  By integrating the scripts with Git, a user can utilize Git workflows to manage source control, versioning, and pipelines.

Setting up the repository can be done following the same steps from [Export Templates](#Export-Templates).  The only differences is the `provides` selection changes to `jobs`.

Jobs need to be defined in `/jobs/` directory at the root of a Git repository.

An example tree for `/jobs/`.

```
▶ tree jobs 
jobs
├── __init__.py
└── get-device-connection.py

1 directory, 2 files
```

Once the repository is created in Nautobot.
![](./images/git-data-source_10.png)

> > Note: The same repository and branch can be used for the different `provides` methods.  Nautobot Git as a data source will look for specific root directory names.

Reviewer 3: ```suggestion
!!! tip
    The same repository and branch can be used for the different `provides` methods.  Nautobot Git as a data source will look for specific root directory names.
```

Once the scripts have been pushed into the repository, a sync needs to be executed, after which navigating to Jobs via **Extensibility -> Jobs** will show the new jobs loaded from the Git repository.

> ![](./images/git-data-source_11.png)

Reviewer 3: This screenshot is pretty wide and involves a pretty significant amount of whitespace. Consider making the browser window narrower and retaking the screenshot?

Jobs now shows the job from the Git repository.

> ![](./images/git-data-source_12.png)

Reviewer 3: Same comment about this screenshot, as well as images 4 through 9 and 14 through 18.

At this point all changes, and history can be kept using Git.  A simple `sync` operation can be done from Nautobot to pulldown any changes.

### Config Contexts

> Detailed information on [config contexts](https://nautobot.readthedocs.io/en/latest/models/extras/gitrepository/#configuration-contexts).

Reviewer 3: ```suggestion
Detailed information on [config contexts](../models/extras/gitrepository.md#configuration-contexts) in Git Repositories.
```

Config contexts may be provided as JSON or YAML files located in `/config_contexts/`.

> Config contexts can be used to provide additional details to different automation tooling.  For example `Ansible vars`, or any other data that you can't natively store in Nautobot.  It can also be used in the Golden Configuration Nautobot plugin to provide extra details to generate configuration templates.

Reviewer 3: ```suggestion
Config contexts can be used to provide additional details to different automation tooling.  For example Ansible variables, or any other data that you can't natively store in Nautobot.  It can also be used in the Golden Configuration Nautobot plugin to provide extra details to generate configuration templates.
```

# Git as a Data Source

> The "Git™ as a Data Source" feature was developed to provide the ability to populate existing data, templates, scripts, and much more into Nautobot; while leveraging the benefits that tools such as GitHub and GitLab already provide, including issue tracking, discussions, pipelines, and approvals.

Reviewer 4: Any reason you're including TM?

Author: Might be over thinking this one, but read https://git-scm.com/about/trademark section 2.1.  I maybe confusing whether its necessary or not.

> 

Reviewer 4: Would love an example this early on in the user guide.  

...tools such as GitHub and GitLab already provide, including issue tracking, discussions, pipelines, and approvals. For example, imagine being able to have users approve the YAML data that is used for Nautobot _config contexts_ along with running tests on that data, or having users approve Jinja2 templates that are used for Nautobot _export templates_ (just as a few examples).  That's what Git as a Data Source offers.

For more technical details on how to use this feature, please see the documentation on [Git Repositories](../models/extras/gitrepository.md).

## Supported Providers
The feature uses the concept of a `provides` field to map a repository to a use case. A list of the supported options is provided below.

### Core Functionality
|Name|Summary|
|:--|:--|

> |[Export Templates](../models/extras/exporttemplate.md)|Nautobot allows users to define custom templates that can be used when exporting objects.|

Reviewer 4: These links seem to link to markdown files in a repo.  Can you link them appropriately to RTD?

Reviewer 3: They're relative links to other documentation pages in the same directory. RTD/Mkdocs converts them to the appropriate relative hyperlinks when rendering to HTML - see https://nautobot--184.org.readthedocs.build/en/184/user-guides/git-data-source/#core-functionality for examples

|[Jobs](../additional-features/jobs.md)|Jobs are a way for users to execute custom logic on demand from within the Nautobot UI. Jobs can interact directly with Nautobot data to accomplish various data creation, modification, and validation tasks.|
|[Config Contexts](../models/extras/configcontext.md)|Config contexts can be used to provide additional data that you can't natively store in Nautobot.|

### Examples of Plugins Defining Additional Providers

> |Name|Summary|Related Plugin|

Reviewer 4: Right above the table, can you add a one-line description being slightly more verbose about what this means and the value of it for plugins?

|:--|:--|:--|
|Backup Configs|Backup configuration data.|[Golden Config](https://github.com/nautobot/nautobot-plugin-golden-config)|
|Intended Configs|Stores the intended configurations, this grabs Nautobot data and runs through Jinja Templates.|[Golden Config](https://github.com/nautobot/nautobot-plugin-golden-config)|
|Jinja Templates|Repository that holds Jinja templates to be used to generate intended configs.|[Golden Config](https://github.com/nautobot/nautobot-plugin-golden-config)|

## Repository Details

> 

Reviewer 4: Same thing here.  What are "Repository Details" ? Why is it important at this point in the article?  Would add an introduction, transition, or move it.

|Field|Explanation|
|:--|:--|
|Name|User friendly name for the repo.|
|Slug|Computer-friendly name for the repo. Auto-generated based on the `name` provided, but you can change it if you wish.|
|Remote URL|The URL pointing to the Git repo. Current git url usage is limited to `http` or `https`.|
|Branch|The branch in the Git repo to use. Defaults to `main`.|
|Token|The token is a personal access token for the `username` provided.  For more information on generating a personal access token see the corresponding links below. **Token is the only option supported currently.**|
|Username|The Git username that corresponds with the personal access token above. Note not required for GitHub, but is for GitLab.|
|Provides|Resource type(s) provided by this Git repo.|

- [GitHub Personal Access Token](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)
- [GitLab Personal Access Token](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html)
- [Bitbucket Personal Access Token](https://confluence.atlassian.com/bitbucketserver/personal-access-tokens-939515499.html)

## Using Git Data Sources

> This section will focus on examples and use the `user-guide` branch on the demo-git-datasources repo: `https://github.com/nautobot/demo-git-datasource`.

Reviewer 4: add backticks around `demo-git-datasources`  (that has an "s" and the URL doesn't).  Please correct the one that needs correcting :)

### Export Templates

[Export Templates](../models/extras/exporttemplate.md) allow a user to export Nautobot objects based on a custom template.  Export templates can change over time depending on the needs of a user.  Allowing export templates to reference a Git repo makes managing templates easier.

A template can be used to put objects into a specific format for ingestion into another system, tool, or report.  It is possible that different templates are needed depending on specific users or teams.  This can lead to sprawl of export templates.  To keep accurate templates synced with Nautobot the Git Data Sources extensibility feature can be used.

#### Step 1
Navigate to the Data Sources Git integration. **Extensibility -> Git Repositories**.

> ![](./images/git-data-source_1.png)

Reviewer 3: Best practice for accessibility would be to include appropriate alt text for each of these images, e.g.:

```suggestion
![Menu showing "Git Repositories" item](./images/git-data-source_1.png)
```

Click [+] or [Add]

That loads a default page to add a repository.

> ![Add a New Git Repository](./images/git-data-source_2.png)

Reviewer 4: Would you mind creating a directory for this user guide? So we have:  `./images/git-as-data-source/01-image-name.png`. For each. Also add the 00 at the front of the filename.

!!! note
    By default only config contexts, export templates, and jobs are available resource types.  Others may be added when specific plugins are used.

#### Step 2
Fill out the details for the Git repository. More information on the inputs can be found in the [fields section](#repository-details).

![](./images/git-data-source_3.png)

As soon as you click on **Create**, Nautobot will clone and sync the repository and provide status of the job.

![](./images/git-data-source_4.png)

> ![](./images/git-data-source_6.png)

Reviewer 3: Looking at the new, narrower versions of these images, we *really* should fix how we render timestamps when showing JobResult logs:

![image](https://user-images.githubusercontent.com/5603551/112547424-1c546780-8d91-11eb-9610-54bb05c62277.png)

The repository will now be displayed on the main Git Repository page.

![](./images/git-data-source_5.png)

Once the repository is synced each template will now be available in the Export Templates section.  **Extensibility -> Export Templates**.

![](./images/git-data-source_7.png)

> !!! note
    If the templates don't populate, make sure the Git directory is named `export_templates` and the sub-directory and sub-sub-directory names correctly match the Nautobot `content type`.

Example below:

> ```

Reviewer 3: ```suggestion
```no-highlight
```

▶ tree export_templates 
export_templates
└── dcim
    └── device
        ├── markdown_export.md
        ├── text_export.txt
        └── yaml_export.yml

2 directories, 3 files
```

#### Step 3
Now that the export templates have been loaded into Nautobot they can be utilized as normal.  For example navigate to **Devices -> Devices** and click on **Export** in the top right corner, the dropdown will now include the templates loaded from the Git repository.

The power of having export templates utilizing the Git integration comes with the native source control features that Git comes with.  To illustrate a simple Git sync within Nautobot assume the following template needs to be updated.

Filename: `/export_templates/dcim/device/yaml_export.yml`

Current contents:

> ```jinja

Reviewer 3: FWIW, to enable `jinja` syntax highlighting, I think you'll need to add it in `mkdocs.yml`:

```yaml
    hljs_languages:
        - jinja
        - yaml
```

Author: @glennmatthews  Should we do this? Or just use standard code block?  I don't think the Jinja2 highlighting is a requirement here.

Reviewer 3: I'm in favor of it - it makes the readability a bit better.

    Navigate to **Git Repositories** and select the repository in question.  Once you're in the detailed view you can look through the **Synchronization Status** or **Change Log** tabs.

Now that the Git repository is linked for export templates it can be controlled via the normal Git operations workflow, which allows users or groups of users to perform code reviews using Pull Requests etc.

### Jobs
Jobs are a way for users to execute custom logic on demand from within the Nautobot UI. Jobs can interact directly with Nautobot data to accomplish various data creation, modification, and validation tasks.

For technical details on jobs, please see the documentation on [jobs](../additional-features/jobs.md#jobs).

Jobs allow a user to write scripts in Python.  By integrating the scripts with Git, a user can utilize Git workflows to manage source control, versioning, and pipelines.

Setting up the repository can be done following the same steps from [Export Templates](#Export-Templates).  The only differences is the `provides` selection changes to `jobs`.

Jobs need to be defined in `/jobs/` directory at the root of a Git repository.

An example tree for `/jobs/`.

> ```

Reviewer 3: ```suggestion
```no-highlight
```

### Config Contexts
Detailed information on [config contexts](../models/extras/gitrepository.md#configuration-contexts) in Git Repositories.

Config contexts may be provided as JSON or YAML files located in `/config_contexts/`.

Config contexts can be used to provide additional details to different automation tooling.  For example Ansible variables, or any other data that you can't natively store in Nautobot.  It can also be used in the Golden Configuration Nautobot plugin to provide extra details to generate configuration templates.

A few simple examples of Configuration Context data might be:
 - DNS Servers
 - NTP Servers
 - ACL Data
 - Routing Information such as BGP ASNs etc.

Similar to the other data sources, the repository can be added by navigating to **Extensibility -> Git repositories**. Click on **Add**, and fill out the repository details.

![](./images/git-data-source_13.png)

Once the repository syncs the details can be found in the **Synchronization Status** tab.  For example the platform specifics were synced.

![](./images/git-data-source_14.png)

The repository structure is:

> ```

Reviewer 3: ```suggestion

```no-highlight
```

│   ├── site-a-spine-02.yml
│   ├── site-b-bb-01.yml
│   ├── site-b-leaf-01.yml
│   ├── site-b-leaf-02.yml
│   ├── site-b-rtr-01.yml
│   ├── site-b-rtr-02.yml
│   ├── site-b-spine-01.yml
│   └── site-b-spine-02.yml
├── platform_eos.yml
├── platform_junos.yml
├── platform_nxos.yml
└── role_spine.yml

1 directory, 16 files
```

Configuration Context details:
- Follows an inheritance methodology similar to what Ansible implements.  Global contexts can be overwritten by local contexts at both a group level, as well as at a device specific level.
- Nautobot UI provides a simple view to see merged config contexts.  It can be visualized by navigating to a device and clicking on the **config contexts** tab.

Here's an example, with some of the details omitted for brevity.

![](./images/git-data-source_15.png)

There is a huge benefit to having `config contexts` managed by a Git workflow.  This type of data can be modified often, especially platform specifics, or new device roles.  Utilizing a standard Git workflow allows for all the proper reviews and approvals to be accomplished before accepting the changes into Nautobot for use.

## Additional Git Data Sources

> As seen in [Step 2](#Step-2), the standard installation of Nautobot will come natively with export templates, jobs, and config contexts.  Additional data sources can be incorporated using the Nautobot plugin system.  For example, the [nautobot-plugin-golden-config](https://github.com/nautobot/nautobot-plugin-golden-config) plugin implements four additional data sources.

Reviewer 3: ```suggestion
As seen in [Step 2](#step-2), the standard installation of Nautobot will come natively with export templates, jobs, and config contexts.  Additional data sources can be incorporated using the Nautobot plugin system.  For example, the [nautobot-plugin-golden-config](https://github.com/nautobot/nautobot-plugin-golden-config) plugin implements four additional data sources.
```

Though it might be better to retitle "Step 1" through "Step 3" into something more specific and meaningful?

> As seen in [Fill out Repository Details](#Fill-out-Repository-Details), the standard installation of Nautobot will come natively with export templates, jobs, and config contexts.  Additional data sources can be incorporated using the Nautobot plugin system.  For example, the [nautobot-plugin-golden-config](https://github.com/nautobot/nautobot-plugin-golden-config) plugin implements four additional data sources.

Reviewer 3: Autogenerated section anchors are all-lower-case and case-sensitive.

```suggestion
As seen in [Fill out Repository Details](#fill-out-repository-details), the standard installation of Nautobot will come natively with export templates, jobs, and config contexts.  Additional data sources can be incorporated using the Nautobot plugin system.  For example, the [nautobot-plugin-golden-config](https://github.com/nautobot/nautobot-plugin-golden-config) plugin implements four additional data sources.
```

- Config Contexts
- Backup Configs
- Intended Configs
- Jinja Templates

For more information for the Golden Configuration specific data sources, navigate to [Nautobot Golden Config Repo](https://github.com/nautobot/nautobot-plugin-golden-config/blob/develop/docs/navigating-golden.md#git-settings).

## Common Issues and Troubleshooting
1. Repository is linked, but data is not properly loaded into Nautobot.
    - Validate the root directory is set to the proper name.
    - Export Templates -> `export_templates`.
    - Jobs -> `jobs`.
    - Config Contexts -> `config_contexts`.
2. Synchronization Status Failures.
    - Validate branch is correct and exists in the remote repository.
    ![](./images/git-data-source_16.png)
    - Validate remote url is correct and is the `https` url.  `ssh` urls are not currently supported.

>     ![](./images/git-data-source_17.png)    

Reviewer 3: This image is showing a "Could not read Username" error - is that the correct intended image for this item?

Author: it was the wrong image, great catch.

|Field|Explanation|
|:--|:--|
|Name|User friendly name for the repo.|
|Slug|Computer-friendly name for the repo. Auto-generated based on the `name` provided, but you can change it if you wish.|
|Remote URL|The URL pointing to the Git repo. Current git url usage is limited to `http` or `https`.|
|Branch|The branch in the Git repo to use. Defaults to `main`.|
|Token|The token is a personal access token for the `username` provided.  For more information on generating a personal access token see the corresponding links below. **Token is the only option supported currently.**|
|Username|The Git username that corresponds with the personal access token above. Note not required for GitHub, but is for GitLab.|
|Provides|Resource type(s) provided by this Git repo.|

- [GitHub Personal Access Token](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token)
- [GitLab Personal Access Token](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html)
- [Bitbucket Personal Access Token](https://confluence.atlassian.com/bitbucketserver/personal-access-tokens-939515499.html)

## Using Git Data Sources

> This section will focus on examples and use the `user-guide` branch on the `demo-git-datasources` repo: `https://github.com/nautobot/demo-git-datasource`.

Reviewer: Please change this to link to the branch directly: https://github.com/nautobot/demo-git-datasource/tree/user-guide

### Export Templates

[Export Templates](../models/extras/exporttemplate.md) allow a user to export Nautobot objects based on a custom template.  Export templates can change over time depending on the needs of a user.  Allowing export templates to reference a Git repo makes managing templates easier.

A template can be used to put objects into a specific format for ingestion into another system, tool, or report.  It is possible that different templates are needed depending on specific users or teams.  This can lead to sprawl of export templates.  To keep accurate templates synced with Nautobot the Git Data Sources extensibility feature can be used.

#### Add a Repository
Navigate to the Data Sources Git integration. **Extensibility -> Git Repositories**.

> ![Menu showing "Git Repositories" Item](./images/git-as-data-source/00-git-data-source_1.png)

Reviewer: Please add a height/width on this image. It is very large.

From the default Git repositories view:
![Sync Repository from Default Menu](./images/git-as-data-source/00-git-data-source_8.png)

From the detailed view:
![Sync Repository from Detailed Menu](./images/git-as-data-source/00-git-data-source_9.png)

!!! tip
    Once the repository has been synced it's easy to check the history for the templates.
    Navigate to **Git Repositories** and select the repository in question.  Once you're in the detailed view you can look through the **Synchronization Status** or **Change Log** tabs.

Now that the Git repository is linked for export templates it can be controlled via the normal Git operations workflow, which allows users or groups of users to perform code reviews using Pull Requests etc.

### Jobs
Jobs are a way for users to execute custom logic on demand from within the Nautobot UI. Jobs can interact directly with Nautobot data to accomplish various data creation, modification, and validation tasks.

> For technical details on jobs, please see the documentation on [jobs](../additional-features/jobs.md#jobs).

Reviewer: ```suggestion
For technical details on jobs, please see the [documentation on Jobs](../additional-features/jobs.md#jobs).
```

Once the scripts have been pushed into the repository, a sync needs to be executed, after which navigating to Jobs via **Extensibility -> Jobs** will show the new jobs loaded from the Git repository.

![Default Repository Menu](./images/git-as-data-source/00-git-data-source_11.png)

Jobs now shows the job from the Git repository.

![List of Loaded Jobs](./images/git-as-data-source/00-git-data-source_12.png)

At this point all changes, and history can be kept using Git.  A simple `sync` operation can be done from Nautobot to pulldown any changes.

### Config Contexts
Detailed information on [config contexts](../models/extras/gitrepository.md#configuration-contexts) in Git Repositories.

Config contexts may be provided as JSON or YAML files located in `/config_contexts/`.

Config contexts can be used to provide additional details to different automation tooling.  For example Ansible variables, or any other data that you can't natively store in Nautobot.  It can also be used in the Golden Configuration Nautobot plugin to provide extra details to generate configuration templates.

> A few simple examples of Configuration Context data might be:
>  - DNS Servers
>  - NTP Servers
>  - ACL Data
>  - Routing Information such as BGP ASNs etc.

Reviewer: These aren't rendering correctly. Here's a fix:
```suggestion
A few simple examples of Configuration Context data might be:

- DNS Servers
- NTP Servers
- ACL Data
- Routing Information such as BGP ASNs etc.
```

Similar to the other data sources, the repository can be added by navigating to **Extensibility -> Git repositories**. Click on **Add**, and fill out the repository details.

![Example Details Config Contexts](./images/git-as-data-source/00-git-data-source_13.png)

> Once the repository syncs the details can be found in the **Synchronization Status** tab.  For example the platform specifics were synced.

Reviewer: ```suggestion
Once the repository syncs the details can be found in the **Synchronization Status** tab.  For example, the platform specifics were synced:
```

![Syncronization Menu With Loaded Contexts](./images/git-as-data-source/00-git-data-source_14.png)

The repository structure is:
```no-highlight
▶ tree config_contexts 
config_contexts
├── devices
│   ├── site-a-bb-01.yml
│   ├── site-a-rtr-01.yml
│   ├── site-a-rtr-02.yml
│   ├── site-a-spine-01.yml
│   ├── site-a-spine-02.yml
│   ├── site-b-bb-01.yml
│   ├── site-b-leaf-01.yml
│   ├── site-b-leaf-02.yml
│   ├── site-b-rtr-01.yml
│   ├── site-b-rtr-02.yml
│   ├── site-b-spine-01.yml
│   └── site-b-spine-02.yml
├── platform_eos.yml
├── platform_junos.yml
├── platform_nxos.yml
└── role_spine.yml

1 directory, 16 files
```

> Configuration Context details:
> - Follows an inheritance methodology similar to what Ansible implements.  Global contexts can be overwritten by local contexts at both a group level, as well as at a device specific level.
> - Nautobot UI provides a simple view to see merged config contexts.  It can be visualized by navigating to a device and clicking on the **config contexts** tab.

Reviewer: Lists not rendering properly here, too:
```suggestion
Configuration Context details:

- Follows an inheritance methodology similar to what Ansible implements.  Global contexts can be overwritten by local contexts at both a group level, as well as at a device specific level.
- Nautobot UI provides a simple view to see merged config contexts.  It can be visualized by navigating to a device and clicking on the **config contexts** tab.
```

- [GitLab Personal Access Token](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html)
- [Bitbucket Personal Access Token](https://confluence.atlassian.com/bitbucketserver/personal-access-tokens-939515499.html)

## Using Git Data Sources
This section will focus on examples and use the `user-guide` branch on the `demo-git-datasources` repo: `https://github.com/nautobot/demo-git-datasource/tree/user-guide`.

### Export Templates

[Export Templates](../models/extras/exporttemplate.md) allow a user to export Nautobot objects based on a custom template.  Export templates can change over time depending on the needs of a user.  Allowing export templates to reference a Git repo makes managing templates easier.

A template can be used to put objects into a specific format for ingestion into another system, tool, or report.  It is possible that different templates are needed depending on specific users or teams.  This can lead to sprawl of export templates.  To keep accurate templates synced with Nautobot the Git Data Sources extensibility feature can be used.

#### Add a Repository
Navigate to the Data Sources Git integration. **Extensibility -> Git Repositories**.

> ![Menu showing "Git Repositories" Item](./images/git-as-data-source/00-git-data-source_1.png)

Reviewer 3: I don't think `00-git-data-source_1.png`, `00-git-data-source_2.png`, etc. was what @jedelman8 had in mind for the image names here - I'm guessing he meant `01-git-data-source.png`, `02-git-data-source.png`, etc.

Author: Let me know which is preferred.  I was assuming prepended number was for a versioning technique.  If I need to change them I will.

Author: @glennmatthews  @jathanism  
Fixed, sorry for the confusion.

![](./images/git-data-source_7.png)

> > !!! note

Reviewer 3: I think I missed this in my previous suggested edit, sorry :-)

```suggestion
!!! note
```
